### PR TITLE
[Model] Refactor Gemma4 vision tower with vLLM-native modules

### PR DESCRIPTION
**Pull request by linitra24**
This PR refactors the Gemma4 multimodal vision tower to use vLLM-native modules instead of directly relying on the Transformers vision tower implementation.

The main goal is to make Gemma4 vision tower layers visible to vLLM’s module system, which is a prerequisite for future vision tower LoRA support. This work is related to #42662.

## Notes

The implementation does not use `MMEncoderAttention` yet. Gemma4 image inputs are currently represented as padded patch sequences with a per-token bool padding mask from `pixel_position_ids == -1`. `MMEncoderAttention` is better suited for dense or packed vision sequences using `cu_seqlens`, so using it correctly would require an additional pack/scatter path for valid patch tokens. For this refactor, the attention path keeps the padded layout and uses SDPA with the existing bool mask.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/models/gemma4_mm.py [hunk 1]
```diff
@@ -849,6 +853,583 @@ def get_replacement_audio(item_idx: int):
...
+            pixel_position_ids=pixel_position_ids,
+            padding_positions=padding_positions,
+            output_length=output_length,
+        )
+        hidden_states = hidden_states[pooler_mask]
+
+        if self.config.standardize:
+            hidden_states = (hidden_states - self.std_bias) * self.std_scale
+
+        return BaseModelOutputWithPast(last_hidden_state=hidden_states)
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py]:
```suggestion
        return hidden_states
```

File: vllm/model_executor/models/gemma4_mm.py [hunk 2]
```diff
@@ -849,6 +853,583 @@ def get_replacement_audio(item_idx: int):
...
+        key_states = _apply_multidimensional_rope(
+            key_states, cos, sin, position_ids
+        ).transpose(1, 2)
+
+        value_states = self.v_proj(hidden_states).view(hidden_shape)
+        value_states = self.v_norm(value_states).transpose(1, 2)
+
+        attn_output = _gemma4_vision_attention_forward(
+            self, query_states, key_states, value_states, attention_mask
+        )
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py:1252]: Use `MMEncoderAttention` interface instead.

File: vllm/model_executor/models/gemma4_mm.py [hunk 3]
```diff
@@ -849,6 +853,583 @@ def get_replacement_audio(item_idx: int):
...
+    output_parts = [
+        _apply_rotary_pos_emb(
+            hidden_states=hidden_parts[k],
+            cos=cos_parts[k],
+            sin=sin_parts[k],
+            unsqueeze_dim=unsqueeze_dim,
+        )
+        for k in range(ndim)
+    ]
+    return torch.cat(output_parts, dim=-1)
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py:1139]: I think we can use `ApplyRotaryEmb`:
https://github.com/vllm-project/vllm/blob/5bb8d2767a2829b56e58c68fa8f380e9e4e2bd3e/vllm/model_executor/layers/rotary_embedding/common.py#L122-L127

File: vllm/model_executor/models/gemma4_mm.py [hunk 4]
```diff
@@ -849,6 +853,583 @@ def get_replacement_audio(item_idx: int):
...
+
+        for encoder_layer in self.layers:
+            hidden_states = encoder_layer(
+                hidden_states=hidden_states,
+                position_embeddings=position_embeddings,
+                attention_mask=attention_mask,
+                position_ids=pixel_position_ids,
+            )
+
+        return BaseModelOutputWithPast(last_hidden_state=hidden_states)
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py]:
```suggestion
        return last_hidden_states
```

File: vllm/model_executor/models/gemma4_mm.py (hunk 2 above)
Comment (linitra24) [reply vllm/model_executor/models/gemma4_mm.py:1252]: For MMEncoderAttention, I looked into it, but Gemma4 vision currently relies on a per-sample bool padding mask derived from pixel_position_ids == -1. MMEncoderAttention is built around packed/dense encoder attention with cu_seqlens and does not directly support this arbitrary padding mask. Converting it correctly would require an additional pack/scatter path for valid patch tokens, so I kept the SDPA path in this PR and would prefer to handle the MMEncoderAttention conversion as a follow-up!

File: vllm/model_executor/models/gemma4_mm.py (hunk 3 above)
Comment (linitra24) [reply vllm/model_executor/models/gemma4_mm.py:1139]: Thanks for the suggestion. I looked into using ApplyRotaryEmb here, but kept the HF-style multidimensional RoPE helper for now.

This is similar to the MMEncoderAttention case: Gemma4 vision currently uses batch-specific 2D pixel_position_ids with padded image patches, while the ApplyRotaryEmb path expects a more standard/packed layout. Switching this correctly would require additional layout changes, which I think is better handled as a follow-up.

For this PR, I kept the RoPE behavior aligned with Transformers and focused on moving the vision tower to vLLM-native modules for future LoRA support.

File: vllm/model_executor/models/gemma4_mm.py (hunk 4 above)
Comment (linitra24) [reply vllm/model_executor/models/gemma4_mm.py]: I've addressed this.

File: vllm/model_executor/models/gemma4_mm.py (hunk 1 above)
Comment (linitra24) [reply vllm/model_executor/models/gemma4_mm.py]: I've addressed this.
